### DISCUSSION TITLE
Enhance USB support for RP2 port

### DISCUSSION THREAD

**Discussion by hoihu** [micropython/micropython#11283] [RP2040 / Pico] [answered]
I'd like to have a go on improving the USB support on the `rp2` port. 
My goal would be to improve it so that it provides a similar user experience as the `pyb. USB_VCP` or `pyb.usb_mode` functions/classes.

Intermediate milestones could be:
- add multiple CDC interfaces, make them accessible via `USB_VCP`.. perhaps below `rp2.USB_VCP` instead of `pyb`?
- add dupterm support
- add dynamically selectable interfaces similar as in `pyb.usb_mode` 
- add more alternative interfaces (HID etc).
- ..

I started doing some stuff, managed to have multiple CDC interfaces visible on USB. But now it's time to decide on the setup of the API bindings, specifically how the module should be named, etc. Maybe other things that needs considerations? 

I thought I'd drop this here so that I can align early on the work. Or perhaps somebody else is already doing similar things 😄 

Thanks 
Martin

**Comment by mattytrentini** [comment]
@projectgus has been working on similar USB improvements for some time now - it'd be great if you could help test and provide feedback! See his work in #9497 and [micropython-lib#558](https://github.com/micropython/micropython-lib/pull/558).

**Comment by projectgus** [comment, marked as answer]
Hi @hoihu!

Yes, the long term goal is very much to allow the kinds of thing you've mentioned and to (eventually) have them in a cross-port API, although rp2 is where it's starting.

The basic building blocks are in place in those PRs but I've been stalled on this for a while due to other commitments. As Matt says - testing, feedback, or working on some of the driver implementations or integrations would be amazing! Let me know what you think.

The class driver work I've done so far has been basic HID, and then implementing MIDI as it looked simple (turns out MIDI is really not, haha!) But a CDC class driver (in Python) is planned as well, if that's of interest.

**Reply by ayjaym** [reply]
This  would be incredibly good to have, as I've just discovered to my great disappointment that the Hairless MIDI to serial project binary download links are broken and that I would have to rebuild it from source, and that this appears to be about the only way to currently get MIDI over USB working with Micropython.
I've just completed a cool hardware project with a Pico Pi which is intended to be a python-controllable MIDI control surface. It is fully 3d printable and has an OLED display, rotary encoder, micro sd card reader and a 128 button matrix with 9 additional control buttons, each backed with an addressable RGB led. It also has external MIDI out (i.e old-school MIDI to an external MIDI device).Total build cost is estimated at around £70.
I'm going to make the whole thing fully open source, hardware (and FreeCad design files) and software, and the hardware's working perfectly, but now I've hit these roadblocks with Micropython and USB MIDI which are just incredibly disappointing. CircuitPython wouldn't really be an option, I want the multithreading so that one of the Pico's cores can run stuff in the background and raise events. 
If anyone knows where I can get working binaries for Hairless that would at least let me press on, otherwise, I'm totally stalled on the project, I can't really release it without USB MIDI working.

**Reply by ayjaym** [reply]
![img](https://user-images.githubusercontent.com/43778817/233625656-06d512bb-e685-49b4-83fa-aebd5ae5c909.jpg)

pic of the project running a test program to verify all controls are operating correctly

**Reply by projectgus** [reply]
> If anyone knows where I can get working binaries for Hairless that would at least let me press on, otherwise, I'm totally stalled on the project, I can't really release it without USB MIDI working.

Haha, extreme irony that "Hairless MIDI Serial" was my project as well. I haven't really worked on that for a decade though, and part of my reasoning was that USB MIDI microcontroller options had come a long way since 2011, so it wasn't as necessary!

I'll make sure to ping you once the PR is up to the point of easily being a MIDI device, it shouldn't be too much more work I just need the time to dedicate to it. :grin:

**Comment by hoihu** [comment]
sorry for the delayed reply, 

@mattytrentini thanks for the pointer!

@projectgus I saw your project and it's very interesting. I definitely want to try it out. Amazing to think that all the descriptor stuff and callbacks can be supplied via python - I guess that opens up a whole new level in USB configurability (without recompiling, C code fumbling etc). It's also much better for debugging I presume.. Looks very clean and well documented. Might be also an option for other platforms too.

Maybe handling USB low level in python has some performance impacts? But it could well be that this doesn't matter much for a lot of USB devices.

As for collaboration I'm willing, but can't promise anything since I'm (as you..) also busy with other staff. I'm on macOS so any testing would be around that platform. 

Still I want to try it out! My next steps are to use your PR and make the example work, then continue with adding a new CDC interface (hopefully)

**Reply by hoihu** [reply]
Ok so I tried it and got the mouse example working on macOS Ventura 13.3 without issues.

**Reply by projectgus** [reply]
That's great to hear. Let me know if you hi any other roadblocks.

> Maybe handling USB low level in python has some performance impacts? But it could well be that this doesn't matter much for a lot of USB devices.

It's a good question. We'll have to wait until we can measure a complete implementation to be sure. At this point my guess is that it will be a measurable impact, but only a significant impact for applications that need to saturate the USB link and also need to do a significant amount of processing on the Python side (or at least using the CPU). But we'll see.

**Reply by hoihu** [reply]
Made some progress. I have now 2 CDC interfaces running on my pico, one with the "static" USB descriptor and one using your python lib :)

I created an interface class for both the CDC data and CDC control interface and added the IAD (Interface association descriptor) in front, as it's done within tinyusb. On Wireshark the two descriptor now look the same and macOS reports 2 interfaces. It's very early days and I did not yet do any tests with data submission etc, but it's promising. 

One thing I noted was that `EP_OUT_FLAG` was inverted? That is, 0x80 should be the IN? I stumbled across this when using wireshark.

**Reply by hoihu** [reply]
here is the link to the cdc class: https://github.com/hoihu/micropython-lib/blob/feat-cdc-python/micropython/usbd/cdc.py

**Comment by raveslave** [comment]
Noted that there is support for USB Host in the official RP2040 repo. Any plans to link this up with micropython?   
would of course be fantastic to also get a host midi class added as well
